hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/SCMCommonPlacementPolicy.java

```diff
@@ -426,4 +451,67 @@ public boolean isValidNode(DatanodeDetails datanodeDetails,
     }
     return false;
   }
+
+  /**
+   * Given a set of replicas of a container which are
+   * neither over underreplicated nor overreplicated,
+   * return a set of replicas to copy to another node to fix misreplication.
+   * @param replicas
+   */
+  @Override
+  public Set<ContainerReplica> replicasToCopyToFixMisreplication(
+         Set<ContainerReplica> replicas) {
+    Map<Node, List<ContainerReplica>> placementGroupReplicaIdMap
+            = replicas.stream().collect(Collectors.groupingBy(replica ->
+            this.getPlacementGroup(replica.getDatanodeDetails())));
+
+    int totalNumberOfReplicas = replicas.size();
+    int requiredNumberOfPlacementGroups =
+            getRequiredRackCount(totalNumberOfReplicas);
+    int additionalNumberOfRacksRequired = Math.max(
+            requiredNumberOfPlacementGroups - placementGroupReplicaIdMap.size(),
+            0);
+    int replicasPerPlacementGroup =
+            getMaxReplicasPerRack(totalNumberOfReplicas);
+    Set<ContainerReplica> copyReplicaSet = Sets.newHashSet();
+
+    for (List<ContainerReplica> replicaList: placementGroupReplicaIdMap
+            .values()) {
+      if (replicaList.size() > replicasPerPlacementGroup) {
+        List<ContainerReplica> replicasToBeCopied = replicaList.stream()
+                .limit(replicaList.size() - replicasPerPlacementGroup)
+                .collect(Collectors.toList());
+        copyReplicaSet.addAll(replicasToBeCopied);
+        replicaList.removeAll(replicasToBeCopied);
+      }
+    }
+    if (additionalNumberOfRacksRequired > copyReplicaSet.size()) {
```

Comment:
Would this algorithm work:

```
    int replicaCount = replicas.length;
    int remainingRacks = requiredRacks;

    while (orderedRacks.size() > 1 && remainingRacks > 0) {
      int maxOnRack = replicaCount / remainingRacks +
          Math.min(replicaCount % remainingRacks, 1);
      List<String> nodes = orderedRacks.poll();
      if (nodes.size() > maxOnRack) {
        // Copy those over maxPerRack.
        replicaCount = replicaCount - maxOnRack;
      } else {
        // Nothing to remove from this rack
        replicaCount = replicaCount - nodes.size();
      }
      remainingRacks -= 1;
    }
```

Where orderedRacks is the queue you formed, basically we start a grouping of Rack -> List<ContainerReplica> and the number of replicas and the number of racks we require.

Then we check if the biggest rack has over the max it should have, if it does, add the excess to copy.

Then we reduce the replica count by what will remain on this largest rack, and reduce the racks by 1, and do the calculation again to find the next max per rack, and so on. We are sort of recursively peeling off the largest rack and fixing it and then checking the rest until we get it down to just 2 racks or have met the required rack count.

Reply:
This wouldn't work for the case say we have 5 replicas and 4 racks. Say current placement has:
Rack 1: 2
Rack 2: 2
Rack 3: 1
In this case one of the replica from either Rack 1 or Rack 2 has to be copied.
Max Replicas per rack is 2 in this case (5/4) + (5%4=1) = 2
So that is why I added another check for additionalNumberOfRacksRequired & I am doing the same algorithm, but now comparing the rack with most replicas with the rack with second highest number of replicas

Reply:
The algorithm you are suggesting would also work.

Reply:
@sodonnel 
When there is a lot of skew in replicas per rack this might lead to that kind of situation where maxReplicaPerRack would change significantly. 
I came up with this case consider this is placement with total number of replicas=13. I know this is bizarre case but consider this is the distribution.
5 Racks case: 5,4,2,1,1
In the first iteration:
Max Replicas per rack = 13/5 = 2.6 rounded up to 3
We will move 2 replicas from rack 1.
In the second iteration:
Max Replicas per rack = 8/4 = 2
We will move 2 replicas from rack 2
In the third iteration:
Max Replicas per rack = 4/3 = 1.33 rounded up to 2
No replicas would be moved from rack 3
In the 4th iteration:
Max Replicas per rack = 2/2 =. 1
No replicas would be moved from rack 4
In the 5th iteration:
Max Replicas per rack. = 1/1 = 1
No replicas would be moved from rack 5
Thus we are moving 4 replicas according to this 2 algorithm 2 from rack 1 & 2 from rack 2.
But overall requirement by the policy is to just have maximum 3 replicas per rack = 13/5 = 2.6 rounded up to 3
We could have achieved this by removing 2 replicas from rack 1 & 1 replica from rack 2. In this case we are just moving 3 replicas overall which is lesser than 4. So this algorithm might over copy replicas as per my previous comment.

Reply:
Looking at your algorithm again I misunderstood it. This algorithm should work as the max number of replicas being removed would be just the ceiling of the total number of replicas.

Reply:
I have changed the code with the proposed algorithm.